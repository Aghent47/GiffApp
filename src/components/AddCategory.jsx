import { useState } from "react";
import PropType from 'prop-types';
export const Addcategory = ({ onNewCategory }) => {

  const [inputValue, setInputValue] = useState('');

  const onInputChange = ( event ) => {
    /* console.log(event.target.value); */
    setInputValue( event.target.value );
  };

  const onSubmit = ( event ) => {
    event.preventDefault();

    if( inputValue.trim().length <= 1 )return;
    
    onNewCategory( inputValue.trim() );
    setInputValue('');
    
  }

  return (
     <form onSubmit={ onSubmit } aria-label="form" >
      <input type="text" placeholder="Buscar Gif" value={ inputValue } onChange={ onInputChange } />
     </form>
  );
}

Addcategory.PropType = {
  onNewCategory: PropType.func.isRequired,
}