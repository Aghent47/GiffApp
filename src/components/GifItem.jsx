import React, { useState } from 'react';
import { motion } from 'framer-motion';
import { Heart, Download, Share2, Eye, Play } from 'lucide-react';
import PropTypes from 'prop-types';

export const GifItem = ({ title, url, id }) => {
  const [isHovered, setIsHovered] = useState(false);
  const [isLiked, setIsLiked] = useState(false);
  const [imageLoaded, setImageLoaded] = useState(false);

  const handleLike = (e) => {
    e.stopPropagation();
    setIsLiked(!isLiked);
  };

  const handleDownload = (e) => {
    e.stopPropagation();
    // In a real app, you would implement download functionality
    console.log('Downloading:', title);
  };

  const handleShare = (e) => {
    e.stopPropagation();
    // In a real app, you would implement share functionality
    if (navigator.share) {
      navigator.share({
        title: title,
        url: url
      });
    } else {
      navigator.clipboard.writeText(url);
    }
  };

  const cardVariants = {
    hidden: { 
      opacity: 0, 
      scale: 0.8,
      y: 20
    },
    visible: {
      opacity: 1,
      scale: 1,
      y: 0,
      transition: {
        duration: 0.5,
        ease: "easeOut"
      }
    },
    hover: {
      scale: 1.02,
      y: -8,
      transition: {
        duration: 0.2,
        ease: "easeOut"
      }
    }
  };

  const overlayVariants = {
    hidden: { opacity: 0 },
    visible: { 
      opacity: 1,
      transition: { duration: 0.2 }
    }
  };

  const buttonVariants = {
    hidden: { 
      opacity: 0, 
      scale: 0.8,
      y: 10
    },
    visible: { 
      opacity: 1, 
      scale: 1,
      y: 0,
      transition: {
        duration: 0.2,
        ease: "easeOut"
      }
    }
  };

  return (
    <motion.div
      className="gif-card"
      variants={cardVariants}
      initial="hidden"
      animate="visible"
      whileHover="hover"
      onHoverStart={() => setIsHovered(true)}
      onHoverEnd={() => setIsHovered(false)}
      layout
    >
      <div className="gif-image-container">
        {!imageLoaded && (
          <motion.div
            className="loading-skeleton"
            style={{
              position: 'absolute',
              top: 0,
              left: 0,
              right: 0,
              bottom: 0,
              zIndex: 1
            }}
            animate={{
              backgroundPosition: ['0% 0%', '100% 0%', '0% 0%']
            }}
            transition={{
              duration: 2,
              repeat: Infinity,
              ease: "linear"
            }}
          />
        )}
        
        <motion.img
          src={url}
          alt={title}
          className="gif-image"
          onLoad={() => setImageLoaded(true)}
          initial={{ opacity: 0 }}
          animate={{ 
            opacity: imageLoaded ? 1 : 0,
            scale: isHovered ? 1.1 : 1
          }}
          transition={{ duration: 0.3 }}
          whileHover={{ scale: 1.1 }}
        />

        {/* Hover Overlay */}
        <motion.div
          className="gif-overlay"
          variants={overlayVariants}
          initial="hidden"
          animate={isHovered ? "visible" : "hidden"}
        >
          <motion.div className="gif-actions">
            <motion.button
              className={`action-button ${isLiked ? 'liked' : ''}`}
              onClick={handleLike}
              variants={buttonVariants}
              whileHover={{ scale: 1.1 }}
              whileTap={{ scale: 0.9 }}
              aria-label={isLiked ? 'Unlike' : 'Like'}
            >
              <Heart 
                size={18} 
                fill={isLiked ? 'currentColor' : 'none'}
              />
            </motion.button>

            <motion.button
              className="action-button"
              onClick={handleDownload}
              variants={buttonVariants}
              whileHover={{ scale: 1.1 }}
              whileTap={{ scale: 0.9 }}
              aria-label="Download"
            >
              <Download size={18} />
            </motion.button>

            <motion.button
              className="action-button"
              onClick={handleShare}
              variants={buttonVariants}
              whileHover={{ scale: 1.1 }}
              whileTap={{ scale: 0.9 }}
              aria-label="Share"
            >
              <Share2 size={18} />
            </motion.button>
          </motion.div>

          <motion.div 
            className="gif-play-indicator"
            variants={buttonVariants}
          >
            <Play size={24} />
          </motion.div>
        </motion.div>

        {/* Loading Indicator */}
        {!imageLoaded && (
          <motion.div
            className="loading-indicator"
            animate={{ rotate: 360 }}
            transition={{
              duration: 1,
              repeat: Infinity,
              ease: "linear"
            }}
          >
            <div className="loading-spinner" />
          </motion.div>
        )}
      </div>

      <motion.div 
        className="gif-title"
        animate={{
          color: isHovered ? 'var(--accent-primary)' : 'var(--text-primary)'
        }}
        transition={{ duration: 0.2 }}
      >
        {title}
      </motion.div>

      {/* Gradient Overlay for better text readability */}
      <motion.div
        className="gif-gradient-overlay"
        variants={overlayVariants}
        initial="hidden"
        animate={isHovered ? "visible" : "hidden"}
      />
    </motion.div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
};
