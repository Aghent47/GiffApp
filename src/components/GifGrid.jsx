import { useEffect, useState } from "react";

import { getGifs } from "../helpers/getGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ({category}) => {

  const [images, setImages] = useState([]);

  const getImages = async () => {
    const newImages = await getGifs(category);
    setImages(newImages);
  }

  useEffect(() => {
    getImages();
  }, [])
  
  return (
   <>
    <h3>{category}</h3>
    <div className="card-grid " >
      { 
        images.map( (image) => (
         <GifItem 
         key={ image.id  }
         { ...image } // expacir todas las propiedades que vengas de (image)
         /> 
        ))
      }
    </div>
   </>
  )
}
