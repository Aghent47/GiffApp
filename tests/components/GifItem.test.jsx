import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Pruebas en el componente <GifItem/>', () => { 

  const title = 'One piece';
  const url = 'http://localhost:3000/one-piece.gif';

  test('Debe mostrar el componente correctamente', () => {
    const { container } = render(<GifItem title={ title } url={ url } />)

    expect(container).toMatchSnapshot();
  });

  test('Debe mostrar la imagen con el URL y el ALT indicado', () => { 
    render(<GifItem title={ title } url={ url } />)

    /* const img = container.querySelector('img');
    expect(img.src).toBe(url);
    expect(img.alt).toBe(title); */
    const {src, alt } = screen.getByRole('img');
    expect(src).toBe(url);
    expect(alt).toBe(title);
   })

   test('debe mostrar el titulo en el componente', () => { 

    render(<GifItem title={ title } url={ url } />)

    expect(screen.getByText( title )).toBeTruthy();
    
  }); 

});