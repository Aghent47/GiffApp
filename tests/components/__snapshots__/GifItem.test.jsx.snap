// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifItem/> Debe mostrar el componente correctamente 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="One piece"
      src="http://localhost:3000/one-piece.gif"
    />
    <p>
      One piece
    </p>
  </div>
</div>
`;
